flink-core/src/main/java/org/apache/flink/configuration/Configuration.java

```diff
@@ -258,12 +270,28 @@ public void addAll(Configuration other, String prefix) {
         bld.append(prefix);
         final int pl = bld.length();
 
-        synchronized (this.confData) {
-            synchronized (other.confData) {
-                for (Map.Entry<String, Object> entry : other.confData.entrySet()) {
-                    bld.setLength(pl);
-                    bld.append(entry.getKey());
-                    this.confData.put(bld.toString(), entry.getValue());
+        Object lock1 = this.confData;
+        Object lock2 = other.confData;
+
+        // Ensure consistent lock sequence
+        if (System.identityHashCode(lock1) < System.identityHashCode(lock2)) {
+            synchronized (lock1) {
```

Comment:
the locking sequence is the other way round in this instance

Reply:
It is always keep the same order of locking objects, so could avoid dead lock.